### PR TITLE
fix(filters): keep gaussian kernel finite for narrow sigma

### LINKED ISSUE #4589: W1b: `RandomGaussianIllumination` returns an all-NaN image for a `sigma` its own constructor admits (0/0 in `filters.kernels.gaussian`)
### Describe the bug

`RandomGaussianIllumination` accepts `sigma=0.0` — its own constructor check is
`KORNIA_CHECK(all(0 <= el <= 1 for el in sigma), "sigma of gaussian value must be between 0 and 1.")`
(`kornia/augmentation/_2d/intensity/gaussian_illumination.py:177-180`) — and then returns an image of
**all NaN**, with no warning.

The sibling Gaussian path guards against exactly this: `gaussian_blur2d` and `RandomGaussianBlur` reject
a zero sigma with `sigma must be positive, got tensor([[0., 0.]])`. Only the illumination generator admits it.

### Root cause

`kornia/filters/kernels.py:119-120`:

```python
gauss = torch.exp(-x.pow(2.0) / (2 * sigma.pow(2.0)))
return gauss / gauss.sum(-1, keepdim=True)
```

At `sigma == 0` every exponent is `-inf` (or `0/0` at the mean), so `gauss.sum(-1)` is `0` and the
normalisation is `0 / 0`.

It is not only the exact point. `GaussianIlluminationGenerator.forward`
(`kornia/augmentation/random_generator/_2d/gaussian_illumination.py:110-125`) scales the drawn sigma by
the axis length (`sigma_x = width * sigma`) and, for an **even**-length axis, `gaussian` adds the `+0.5`
half-pixel offset — so no sample sits at the mean and the whole kernel underflows for a small non-zero
sigma as well:

```
gaussian(window=4, sigma=0.02) -> [nan, nan, nan, nan]
gaussian(window=5, sigma=0.02) -> [0.0, 0.0, 1.0, 0.0, 0.0]
```

### Reproduction steps

```python
import torch
from kornia.augmentation import RandomGaussianIllumination

torch.manual_seed(0)
aug = RandomGaussianIllumination(p=1.0, sigma=0.0)   # admitted by the class's own KORNIA_CHECK
print(aug(torch.rand(1, 3, 4, 4))[0, 0])
```

```
tensor([[nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan],
        [nan, nan, nan, nan]])
```

All four dtypes NaN (`float16`, `bfloat16`, `float32`, `float64`), every image size tried
(`1x1, 2x2, 3x3, 4x4, 8x8, 32x32`), 100% of pixels.

NaN hits out of 30 seeds, `sigma=(s, s)` on an `n x n` image, default `center`, CPU:

| n \ s | 0.0 | 0.001 | 0.005 | 0.01 | 0.02 | 0.05 | 0.1 |
|---|---|---|---|---|---|---|---|
| 1 | 30 | 26 | 26 | 26 | 26 | 26 | 0 |
| 2 | 30 | 30 | 30 | 30 | 0 | 0 | 0 |
| 3 | 30 | 5 | 5 | 5 | 5 | 0 | 0 |
| 4 | 30 | 30 | 30 | 0 | 0 | 0 | 0 |
| 8 | 30 | 30 | 0 | 0 | 0 | 0 | 0 |
| 32 | 30 | 30 | 0 | 0 | 0 | 0 | 0 |

The even rows lose the on-grid sample to the half-pixel offset, which is why `n=2` and `n=4` NaN at a
sigma where `n=3` does not.

The class default `sigma=(0.2, 1.0)` is clean: 200 seeds x sizes 1..4, zero NaN. `RandomLinearIllumination`
and `RandomLinearCornerIllumination` have no Gaussian kernel and are unaffected.

### Expected behavior

Either reject a sigma that cannot produce a normalizable kernel, the way `gaussian_blur2d` already does,
or make `gaussian` fall back to a unit impulse at the mean. A constructor that admits a value and then
silently emits NaN is the worst of the three.

The narrow fix is a positivity guard in `GaussianIlluminationGenerator` mirroring `gaussian_blur2d`'s;
the broader one is to make `kornia.filters.kernels.gaussian` refuse a non-positive sigma, since its
`0 / 0` is reachable from any caller that does not pre-check.

### Environment

kornia at `8aca2838f` (PR #4566 head), torch 2.14.0, Python 3.11.14, macOS arm64, CPU.
Found while reviewing #4566; the batch 6c convention block for this class currently states that the
output "stays inside `[0, 1]` even when the input does not", which this falsifies.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

### PR DESCRIPTION
**Pull request by devtejasx**
## What does this pull request do?

`kornia.filters.kernels.gaussian` divided `exp(-x^2 / 2sigma^2)` by its sum. When sigma is small compared to the distance from the mean to the nearest sample, every term underflows and the division is 0 / 0. At sigma = 0 it is 0 / 0 at the mean too. `RandomGaussianIllumination` accepts `sigma=0` and multiplies the sampled sigma by the image size, so it returned all-NaN images.

The kernel is now computed as a softmax over `-(x^2 - min x^2) / 2sigma^2`. The nearest sample always has logit 0, so the result stays finite and tends to an impulse there (split evenly for an even window). At sigma = 0 the impulse is used directly. For normal sigmas the values are unchanged, up to float rounding.

I fixed this in `gaussian` rather than only rejecting sigma = 0 in the augmentation. As the issue's table shows, small non-zero sigmas NaN as well, and other callers of `gaussian` can hit it without a pre-check.

Changelog fragment: `changelog.d/+gaussian-kernel-underflow.fixed.md` (will rename once the PR number exists).

## Related issue or discussion

Fixes #4589

## What did you check?

- Reproducer from the issue (sizes 1/2/3/4/8/32 x sigma 0 ... 0.1 x 30 seeds): 540 NaN runs before, 0 after.
- New tests `test_gaussian_narrow_sigma` and `TestRandomGaussianIllumination::test_narrow_sigma_not_nan` fail on main (15 failures) and pass with the fix.
- `tests/filters` with `--dtype float32,float64,float16`: the failure list is identical to main except one test that now passes (`test_gaussian[cpu-float16-11-11.0-3.0]`); the other 18 float16 failures are already on main.
- `tests/augmentation` float32+float64: 5552 passed, 0 failed.
- Other callers of the kernel (`tests/losses`, `tests/metrics`, `tests/enhance`, pyramid, elastic transform, affine shape, orientation, SIFT desc): all pass.
- `ruff check` / `ruff format --check` clean.

CPU only, torch 2.13, Windows.

## How did AI help?

I used Claude Code to trace the NaN to the normalisation, draft the log-space version and the tests, and run the suites. I checked the math for the even-window and zero-sigma cases and compared the before/after test results myself.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Review of `9114e6ff`.

Thanks for this — the test quality is good (reverting the production change fails 30 nodes, and five one-token mutations are all caught), and you and the author of #4592 independently picked the same structurally correct mechanism. But #4592 fixes the same issue with strictly fewer side effects, so **I am recommending we close this in favour of #4592.** Four of the problems below are independent of the duplication, so I want them recorded.

### [P1] The fix is partial in float16, while the PR claims `Fixes #4589`

Your reproducer swept `sigma >= 0.001`, which is above the float16 threshold. Below it, I measured on your head:

```
get_gaussian_kernel1d(5, 1e-4, dtype=torch.float16)                            -> NaN   (main: NaN)
gaussian_blur2d(rand(1,1,8,8,dtype=float16), (5,5), (1e-4,1e-4))               -> NaN   (main: NaN)
#4592, same two calls                                                          -> finite
```

float16 `sigma` in `(0, 1.698e-4]` is still all-NaN, at **every** window size — odd windows get no improvement over main at all. `-dist / (2 * sigma.pow(2))` is `0/0` at the nearest tap because `sigma.pow(2)` itself underflows (`float16(1e-4)**2 == 0.0`; the smallest subnormal is `5.96e-08`), and `torch.where(sigma > 0, logits, impulse)` picks that NaN branch since `1e-4 > 0`. Same band on torch 2.5.1. This is reachable through the public API with a *positive* sigma that clears `gaussian_blur2d`'s own positivity guard, so the closing keyword over-claims. #4592's `masked_fill(dist == 0, 0.0)` covers it because it keys on the distance rather than on sigma.

### [P1] `torch.where(sigma > 0, ...)` swallows a NaN sigma

```
get_gaussian_kernel1d(5, float('nan'))
  main / #4592 : propagates NaN
  this head    : [0., 0., 1., 0., 0.]
```

A NaN sigma arriving from upstream now produces a plausible-looking impulse instead of propagating.

### [P1] Negative sigma silently changes meaning

```
get_gaussian_kernel1d(5, -1.0)
  main / #4592 : [0.0545, 0.2442, 0.4026, 0.2442, 0.0545]     (the sigma=1 kernel)
  this head    : [0., 0., 1., 0., 0.]
```

Only `sigma**2` is used, so negative sigma has always meant `|sigma|`; `gaussian_discrete_erf` takes `sigma.abs()` explicitly, so that is the deliberate house convention. `get_gaussian_kernel1d`/`get_gaussian_kernel2d` have no positivity guard, so this is reachable.

### [P1] The `softmax` rewrite moves ordinary kernels, and that turns two green tests red

Normalising through `softmax` reorders the arithmetic even when `dist.amin` is 0, so odd-window default-mean kernels move too — 98/170 bfloat16 configs differ, versus 0/170 for #4592's distance-shift. On current main merged with your head:

```
$ pytest tests/filters tests/feature/test_responces_local_features.py -q --device=cpu --dtype=bfloat16
+ FAILED tests/feature/test_responces_local_features.py::TestCornerGFTT::test_corners_batch[cpu-bfloat16]
+ FAILED tests/filters/test_canny.py::TestCanny::test_magnitude[cpu-bfloat16]
```

Neither is in `testing/half_precision_xfails/cpu_bfloat16.txt`, and both pass on main and on #4592.

### [P2] Four manifest lines become strict XPASS and are not deleted

Your body notes one test flipping to pass; there are four. The bfloat16 CI leg says it directly:

```
ERROR: 3 known half-precision failures need updates in testing/half_precision_xfails/cpu_bfloat16.txt
  tests/filters/test_gaussian.py::test_gaussian[cpu-bfloat16-11-11.0-3.0-expected3]: observed call strict XPASS
  tests/filters/test_gaussian.py::test_gaussian[cpu-bfloat16-11-5.0-None-expected1]: observed call strict XPASS
  tests/filters/test_gaussian.py::test_gaussian[cpu-bfloat16-5-1.0-None-expected0]: observed call strict XPASS
```

plus `cpu_float16.txt:226`. A fix that makes a manifest node pass must delete its line in the same PR.

### [P2] Fragment name

`changelog.d/+gaussian-kernel-underflow.fixed.md` is a slug, which is why `Check changelog fragments` is red — the checker requires `<PR>.(added|fixed|breaking).md` and exempts only `+migration-<number>`. It has to be `4593.fixed.md`.

### [P2] Stale base

You branched from `90650596d`; #4566 has since landed a strict wart pin (`tests/augmentation/test_conventions_intensity_ops.py:1488`) and a `#4589` warning in `kornia/augmentation/_2d/intensity/gaussian_illumination.py` (lines 74 and 92-107). Both need handling in whichever PR lands — noting it for completeness, since #4592 owes the same.

### [P3] `gaussian(0, 1.0)`

`amin(): Expected reduction dim -1 to have non-zero size.` where main returns `[[]]`. Same degenerate-shape point as #4592.

### What you did close

For the record, the augmentation-level symptom is fully fixed here: the issue's exact reproduction is finite, its 7x6 seed table is 0/30 NaN everywhere, my 9-pair `n*sigma` table is 9/9 finite, and float16/bfloat16 are finite down to `sigma=1e-6` — the residual band is not reachable *through that augmentation*, because the generator scales sigma by the axis length. The implementation choice is what separates the two PRs, not the effort.


Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
Closing this in favour of **#4592**, which I am taking as the fix for #4589.

This one was not decided on test quality — your tests are the better of the two, and I said so in my [review above](https://github.com/kornia/kornia/pull/4593#pullrequestreview-5225392503): reverting the production change makes them fail, which is the property I most want and do not always get. It was decided on the production semantics, which I ran side by side in three worktrees.

Three things kept it from being the one that lands:

1. **The float16 NaN is not fully closed.** `sigma ∈ (0, 1.7e-4]` in float16 still comes back all-NaN here, because such a sigma is positive and therefore clears `gaussian_blur2d`'s own guard while still underflowing every sample inside the kernel. That is the exact band #4589 is about.
2. **A NaN sigma silently becomes an impulse** rather than propagating or raising, and **negative-sigma semantics change silently** — both are behaviour changes outside the issue's scope, and neither is documented or pinned.
3. **The `softmax` rewrite turns two currently green bfloat16 tests red.**

For contrast, #4592 is bit-identical to main across all 170 ordinary odd-window configurations, so it changes only the degenerate band. I verified all three variants' semantics myself rather than reading them off the diffs.

If you want to carry the part that is genuinely yours: the test design here is worth more than the production diff. #4592 currently owes a docstring for the new `sigma=0` impulse contract and a pin for the `(0, 1.7e-4]` float16 band specifically — a PR adding the second of those on top of #4592, in the style you used here, would be very welcome. Say the word and I will note it on #4589 so it stays yours.

Thank you for the work.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
